### ISSUE TITLE
Softmax / Categorical sampled logit=-float('inf') that has prob=0

### ISSUE THREAD

**Issue by Kelym** [pytorch/pytorch#10712]
## Issue description

**TL;DR** setting logit[i]=-float('inf') does not prevent torch.nn.functional.softmax from sampling that bit. 

 
I am training a seq2seq agent which will output `logit` then I'd use softmax and categorical to sample from it. Sometimes I will mask out a specific logit by setting `logit[x,y]=-float('inf')`. I assumed that this action will make `m` never sampling an `a` such that `a[x]=y`. However I observed that this is not the case (happened once after 2 hours of training).

## Code example

```
probs = F.softmax(logit,dim=1)
m = D.Categorical(probs)
a = m.sample()
```
 
Printing out the value: 
```
logit[i] = tensor([ 3.1237, -0.6266, -3.3107, -4.1575,    -inf,  0.8783], 
       device='cuda:0', grad_fn=<SelectBackward>)
m.probs[i] = tensor([0.8836, 0.0208, 0.0014, 0.0006, 0.0000, 0.0936], 
       device='cuda:0', grad_fn=<SelectBackward>)
a_t[i]=4
m.log_prob(a_t)[i] = -15.9424
```

## System Info

```
Collecting environment information...
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1060 6GB

Nvidia driver version: 396.26
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip3] numpy (1.14.0)
[pip3] numpydoc (0.7.0)
[conda] cuda90                    1.0                  h6433d27_0    pytorch
[conda] pytorch                   0.4.1           py36_cuda9.0.176_cudnn7.1.2_1    soumith
[conda] torchvision               0.1.9            py36h7584368_1    soumith
```

**Comment by Kelym** [comment]
Ok I realized that we have a logit option in torch.distributions.Categorical now. Hopefully this will avoid the numerical instability.
